### PR TITLE
Use Results class for PCA Analysis

### LINKED ISSUE #3275: update PCA analysis to use new .results
Follow the new AnalysisBase API and store results in pca.PCA in `.results` instead of class attributes.

See omnibus issue #3261 for background and details. See PR #3263 as an example (DensityAnalysis)
- [ ] put all generated data into `self.results` instead of `self`
- [ ] keep managed attributes in place with deprecation message
- [ ] test deprecation messages/old attribute access
- [ ] update CHANGELOG: one entry in Changes (use .results) and one entry in Deprecations (attribute access to be removed in 3.0.0)

### PR DESCRIPTION
**Pull request by PicoCentauri**
Fixes #3275 

Changes made in this Pull Request:
Move the results of the PCA analysis to the new results class

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:314]: We'll need tests for these new warnings.

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:92]:
```suggestion
``results.cumulated_variance``. The value at the ith index of `results.cumulated_variance`
```
?

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:142]: Seems fine to me, but just checking; @lilyminium @orbeckst is this the new style we are now following? (I didn't follow the whole docstring conversation too closely it seems..)

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:168]: Add a `. versionadded:: 2.0.0` here.

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:179]: As above (versionadded)

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:192]: As above (versionadded)

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:202]: As above (versionadded)

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:234]: we'll probably want a `versionchanged` entry here just to say that we are now storing results using `MDAnalysis.analysis.base.Results`.

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:250]: I realise it's not being used anywhere, but this might actually be worth keeping (if only commented off), I think this is part of an unintentionally killed code path that now leads to #2728

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple comments, not too sure about the docstring change (it looks good to me, but I don't remember if that was the direction we wanted to go to or not).

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:314]: Tests should be there

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:92]: Yes!

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/pca.py:314]: Sorry my bad, my eyes completely glossed over `test_pca_attr_warning` for some reason.

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:142]: At least we used this style in #3233

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:168]: Good point!

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:250]: Honestly, I would like to remove or only keep the positions. Looking forward to mda_cli I see no way to save a `AtomGroup` in way tu use it later without launching a python interpreter...

Comment (orbeckst) [reply package/MDAnalysis/analysis/pca.py:250]: There might simply be results that mdacli cannot save out of the box but they are necessary for programmatic downstream analysis. That's just something mdacli needs to deal with separately.

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/pca.py:135]: technically (=NumpyDoc) speaking, the synopsis should not be a code example. However, we're not fixing docs right not so I am more than happy to just have some stylistic changes & sort out docs later.

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/pca.py:146]: We are moving towards documenting classes in the class string instead of `__init__`... IIRC. However, right now I am not overly concerned with where the docs are, as long as information is either in class or init doc string with the correct attributes etc I am happy. 

I support @PicoCentauri 's edit here (and move of examples into class doc strings in other PRs) — but I am not going to hold up any other PR if it doesn't do exactly the same thing. If someone wants to raise (or find...?) and issue to decide on class doc then please do. But it can really wait post 2.0.0.

**Review by orbeckst — APPROVED** [review]
lgtm

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:301]: Ahh make sense. I am fine if we postpone this a bit.

**Review by fiona-naughton — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/pca.py:135]: Starting points for docs is https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation

**Comment by orbeckst** [comment]
Not sure where we are with the soft code freeze. However, if we want PCA for the workshop then this PR needs to go in, even if mean atoms still needs fixing. 

However, we can always add fixes, we have a hard time putting in breaking changes so I would suggest to move forward with this PR. 

I’m sorry, I have limited bandwidth today to do more.

**Comment by IAlibay** [comment]
> Not sure where we are with the soft code freeze. However, if we want PCA for the workshop then this PR needs to go in, even if mean atoms still needs fixing.
> 
> However, we can always add fixes, we have a hard time putting in breaking changes so I would suggest to move forward with this PR.
> 
> I’m sorry, I have limited bandwidth today to do more.

Sorry, I can't do any more today. If @fiona-naughton or one of the @MDAnalysis/coredevs wants to push this through then that'd be great. See discord for more on what our priority merges should be.

**Comment by fiona-naughton** [comment]
If it's agreed that `mean_atoms` should have the average positions as calculated/provided to `self.means`, then that could be fixed by making a dummy atom group and populating the positions with the values from `means` something like I'm currently doing in #3296 (thought there I didn't need to worry about atom typing etc).

But the updated docs here reflects that it's just the atom selection rather than implying anything about positions and it's not used anywhere else, so the only issue I can see is the name being misleading - so ultimately I'd say it's fine to just leave it as is. Eventually I think we can get rid of it, or just return the position array instead, but that can wait for another day.

tl;dr I say leave `mean_atoms` for now and decide what to do with it later; everything else lgtm

File: package/MDAnalysis/analysis/pca.py
Comment (PicoCentauri) [reply package/MDAnalysis/analysis/pca.py:135]: Thanks. I'm would especially like to make consistent docs for classes and will open an issue when 2.0 is out.

**Comment by orbeckst** [comment]
Will look at this PR as soon as PR #3296 is merged.

**Comment by orbeckst** [comment]
Note that PR #3296 removes `mean_atoms`.

**Comment by orbeckst** [comment]
@fiona-naughton @IAlibay please re-review after removal of mean_atoms. Thanks.

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/pca.py:167]: Merged with the 2.0.0 entry below during merge.

**Comment by orbeckst** [comment]
The [docs for MDAnalysis.analysis.pca.PCA.cumulative_overlap](https://mdanalysis--3285.org.readthedocs.build/en/3285/documentation_pages/analysis/pca.html#MDAnalysis.analysis.pca.PCA.cumulative_overlap) and [rmsip](https://mdanalysis--3285.org.readthedocs.build/en/3285/documentation_pages/analysis/pca.html#MDAnalysis.analysis.pca.PCA.rmsip) contain a badly rendered _See Also_ section. Would you be able to fix these as well? I am appealing to your 🇩🇪 sense for tidiness ;-).

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/pca.py:453]: Builds a link to itself (even though labeled func and not meth), needs to be
```suggestion
        :func:`~MDAnalysis.analysis.pca.rmsip`
```

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/pca.py:506]: Builds a link to itself (even though labeled func and not meth), needs to be
```suggestion
        :func:`~MDAnalysis.analysis.pca.cumulative_overlap`
```

**Review by orbeckst — CHANGES_REQUESTED** [review]
incorrect links in See Also ... maybe my suggestions fix it

**Comment by orbeckst** [comment]
@fiona-naughton @IAlibay I assume you're knackered after non-stop workshopping but if you could have a look if this PR finds your approval then we could make another big check mark. (Fixing docs is ongoing)

**Comment by PicoCentauri** [comment]
Thanks, for fixing the link @orbeckst!

File: package/MDAnalysis/analysis/pca.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/pca.py:46]:
```suggestion
explains. Stored in :attr:`PCA.results.cumulated_variance`, a ratio for each number of
```
? (it'll need fixes for the number of words since I think this suggestion is > 79 chars

**Review by IAlibay — APPROVED** [review]
Just the one change, otherwise lgtm (I'll approve so I'm not blocking).

File: package/MDAnalysis/analysis/pca.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/pca.py:46]: You're right, well spotted. I'll fix it without using the suggestion.

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
I am messing up with the docs — I'll now fix this for once and for all.

**Review by fiona-naughton — APPROVED** [review]
lgtm :)

**Comment by orbeckst** [comment]
@IAlibay , ready for you; the docs for rmsip and cumulative_overlap finally show up, too!

**Comment by IAlibay** [comment]
Thanks for all the work here @PicoCentauri and @orbeckst !

**Pull request merged**
